include/openthread/dns.h

```diff
@@ -91,9 +91,51 @@ typedef struct otDnsTxtEntry
     const char *   mKey;
     const uint8_t *mValue;       ///< The TXT record value or already encoded TXT-DATA (depending on `mKey`).
     uint16_t       mValueLength; ///< Number of bytes in `mValue` buffer.
-    uint8_t mKeyLength; ///< Number of bytes in `mKey` buffer. MUST be set even if `mKey` is a null-terminated string.
 } otDnsTxtEntry;
 
+/**
+ * This structure represents an iterator for TXT record entires (key/value pairs).
+ *
+ * The data fields in this structure are intended for use by OpenThread core and caller should not read or change them.
```

Comment:
Instead of having the documentation tell users don't touch this structure, I think it is better to hide the structure definition as we do for many other types (e.g. `otInstance`, `otSrpServerHost`...).

```c++
typedef struct otDnsTxtEntryIterator otDnsTxtEntryIterator;
// or
typedef void otDnsTxtEntryIterator;
```
Both should work well. Thoughts?

Reply:
I don't think it would be possible in this case to hide the definition of the `struct otDnsTxtEntryIterator`.

- In this case, we expect the user to define a variable (allocate an instance) of this iterator type and then pass a pointer to it to us (OT core), so we can populate and use it. 
- In other cases where we use opaque types, the user does not need to allocate an instance of the type, so hiding the definition works perfectly.
  - In the other cases, users only deal with pointers to a type that OT core passes to the user. 
- We can always have a pointer to a type which is not fully defined/known by the complier, but as soon as we need to define a variable/instance of the type, the compiler needs to know the full definition (mainly to determine its size).

Reply:
yes, you are right.
I didn't realize that the iterator must be passed in from the caller.